### PR TITLE
Implement pillars in DB using postgresql JSONB type

### PR DESCRIPTION
**Pull request by cbosdo**
## What does this PR change?

Implements: https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00080-jsonb-pillar-db-storage.md

## GUI diff

No difference.

- [X] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

May be some documentation on how to refresh the pillar of some minions could be nice to help migrating to the new pillar data.

- [X] **DONE**

## Test coverage
- Unit tests were added

- [X] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/15378

- [X] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by cbosdo** [comment]
Salt requirements:

* https://github.com/openSUSE/salt/pull/426
* https://github.com/openSUSE/salt/pull/425
* https://github.com/openSUSE/salt/pull/424
* https://github.com/openSUSE/salt/pull/423

Also requires `hibernate-types` packages in OBS as well as its dependencies: [hibernate-types](https://build.opensuse.org/package/show/home:cbosdonnat:branches:systemsmanagement:Uyuni:Master:Other/hibernate-types) and [hibernate-types-kit](https://build.opensuse.org/package/show/home:cbosdonnat:branches:systemsmanagement:Uyuni:Master:Other/hibernate-types-kit)

File: java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java [hunk 1]
```diff
@@ -56,18 +57,21 @@
...
-        pillar.add("org_id", minion.getOrg().getId());
-
-        pillar.add("contact_method", minion.getContactMethod().getLabel());
-        pillar.add("mgr_server", minion.getChannelHost());
+    public Optional<Pillar> generatePillarData(MinionServer minion) {
+        Pillar pillar = minion.getPillars().stream()
+                .filter(item -> item.getCategory().equals(CATEGORY))
+                .findFirst()
+                .orElse(new Pillar(CATEGORY, new HashMap<>(), minion));
+        pillar.getPillar().put("org_id", minion.getOrg().getId());
```
Comment (aaannz) [inline java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: As this is repeating pattern, wouldn't it make sense to add `put` or rather `add` to the Pillar itself?

```java
    public Pillar add(String category, Object data) {
        this.pillar.put(category, data);
        return this;
    }
```

So the usage would be the same as before?

**Review by aaannz — APPROVED** [review]
I have just one question about usage, but LGTM.

File: java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java (hunk 1 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: I pondered doing that... I'll add it then

File: java/code/src/com/suse/manager/webui/services/pillar/MinionCustomInfoPillarGenerator.java [hunk 2]
```diff
@@ -37,25 +38,30 @@
...
      * Generates pillar data containing CustomInfo information of the passed minion
      * @param minion the minion server
      * @return the SaltPillar containing the pillar data
      */
     @Override
-    public Optional<SaltPillar> generatePillarData(MinionServer minion) {
+    public Optional<Pillar> generatePillarData(MinionServer minion) {
         if (minion.getCustomDataValues().isEmpty()) {
             return Optional.empty();
         }
```
Comment (nadvornik) [inline java/code/src/com/suse/manager/webui/services/pillar/MinionCustomInfoPillarGenerator.java:52]: Deleting all Custom Info values should delete the pillar.
I don't see it handled neither here nor in MinionPillarFileManager.
(before it was handled by deleting the pillar file in MinionPillarFileManager)

File: java/code/src/com/suse/manager/webui/services/pillar/MinionPillarFileManager.java [hunk 3]
```diff
@@ -61,24 +52,14 @@ public MinionPillarFileManager(MinionPillarGenerator minionPillarGeneratorIn) {
      */
     public void updatePillarFile(MinionServer minion) {
         this.minionPillarGenerator.generatePillarData(minion).ifPresentOrElse(
-                (pillar) -> this.saveFileToDisk(pillar, this.minionPillarGenerator.getFilename(minion.getMinionId())),
+                (pillar) -> {
+                    minion.getPillars().add(pillar);
```
Comment (nadvornik) [inline java/code/src/com/suse/manager/webui/services/pillar/MinionPillarFileManager.java:56]: Isn't this redundant? Minion pillar is modified already in generatePillarData.

**Review by nadvornik — CHANGES_REQUESTED** [review]

File: java/code/src/com/suse/manager/webui/services/pillar/MinionPillarFileManager.java (hunk 3 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionPillarFileManager.java:56]: right

File: java/code/src/com/suse/manager/webui/services/pillar/MinionCustomInfoPillarGenerator.java (hunk 2 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionCustomInfoPillarGenerator.java:52]: ACK

File: java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java [hunk 4]
```diff
@@ -56,8 +57,12 @@
...
      */
     @Override
-    public Optional<SaltPillar> generatePillarData(MinionServer minion) {
-        SaltPillar pillar = new SaltPillar();
+    public Optional<Pillar> generatePillarData(MinionServer minion) {
+        Pillar pillar = minion.getPillarByCategory(CATEGORY).orElseGet(() -> {
+            Pillar newPillar = new Pillar(CATEGORY, new HashMap<>(), minion);
+            minion.getPillars().add(newPillar);
+            return newPillar;
+        });
```
Comment (nadvornik) [inline java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: Maybe it should call `pillar.getPillar().clear()` here, to be sure that there are no obsolete entries left.

The same in other Pillar Generators.

Comment (aaannz) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: Isn't this implied by being in `orElseGet` part?

Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: well, either we find the pillar of the category we need to generate and reuse it (no need of cleaning) or e got a newly generated one (and added to the collection) from `orElseGet()`. When reusing we may have existing entries to be cleared: you are right!

Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionGeneralPillarGenerator.java:65]: fixed

**Review by nadvornik — APPROVED** [review]
LGTM

**Pull request merged**